pkg/agent/client.go

```diff
@@ -93,7 +94,10 @@ func (cm *connectionManager) Get(connID int64) (*connContext, bool) {
 func (cm *connectionManager) Delete(connID int64) {
 	cm.mu.Lock()
 	defer cm.mu.Unlock()
-	delete(cm.connections, connID)
+	if _, ok := cm.connections[connID]; ok {
+		metrics.Metrics.EndpointConnectionDec()
```

Comment:
Instead of this .Inc() + .Dec() approach, I would have much rather .Set() the gauge based on the len() of cm.connections. But I think that doesn't work because it would be per-client, and we want an overall count metric that does not have server-id cardinality. clientset.go doesn't take any read locks across clients and I didn't want to add that just for metrics.

Any suggestions are appreciated.

Reply:
Hmm. You could maintain a map of `serverID --> connection count`, and have `metrics.Metrics.SetEndpointConnectionCount(cm.serverID, len(cm.connections))`, which would update the map, take the sum across the map, and set the gauge based on that. Just need to make sure to clean up when clients are deleted.

I'm not sure if that's worth the complexity though...

Reply:
That's nice in terms of avoiding new locks, and I think data accuracy. I am most afraid of the "duplicate client" deleted case.

/hold

Reply:
I added a comment with explanation why I think as-is will be accurate.